common/src/primitives/signed_amount.rs

```diff
@@ -0,0 +1,1030 @@
+    pub const MAX: Self = Self::from_atoms(SignedIntType::MAX);
+
+    pub const fn from_atoms(v: SignedIntType) -> Self {
+        SignedAmount { val: v }
+    }
+
+    pub fn into_atoms(&self) -> SignedIntType {
+        self.val
+    }
+
+    pub fn from_unsigned(amount: Amount) -> Option<Self> {
+        let unsigned_atoms = amount.into_atoms();
+        let atoms: SignedIntType = unsigned_atoms.try_into().ok()?;
+        Some(Self::from_atoms(atoms))
+    }
+
+    pub fn into_unsigned(self) -> Option<Amount> {
+        let atoms = self.val;
+        let unsigned_atoms: super::amount::UnsignedIntType = atoms.try_into().ok()?;
+        Some(Amount::from_atoms(unsigned_atoms))
+    }
+
+    pub fn into_fixedpoint_str(self, decimals: u8) -> String {
+        let amount_str = self.val.abs().to_string();
+        let decimals = decimals as usize;
+        let sign = if self.val < 0 { "-" } else { "" };
+        if amount_str.len() <= decimals {
+            let zeros = "0".repeat(decimals - amount_str.len());
+            let result = sign.to_owned() + "0." + &zeros + &amount_str;
+
+            remove_right_most_zeros_and_decimal_point(result)
+        } else {
+            let ten: SignedIntType = 10;
+            let unit = ten.pow(decimals as u32);
+            let whole = self.val.abs() / unit;
+            let fraction = self.val.abs() % unit;
+            let result = format!("{sign}{whole}.{fraction:00$}", decimals as usize);
+
+            remove_right_most_zeros_and_decimal_point(result)
+        }
+    }
+
+    pub fn from_fixedpoint_str(amount_str: &str, decimals: u8) -> Option<Self> {
+        let decimals = decimals as usize;
+        let amount_str = amount_str.trim_matches(' '); // trim spaces
+        let amount_str = amount_str.replace('_', "");
+
+        // empty not allowed
+        if amount_str.is_empty() {
+            return None;
+        }
+        // too long
+        if amount_str.len() > 100 {
+            return None;
+        }
+        // must be only numbers or decimal point
+        if !amount_str.chars().all(|c| char::is_numeric(c) || c == '.' || c == '-') {
+            return None;
+        }
+        // if the amount is only one minus sign, it's invalid
+        if amount_str == "-" {
+            return None;
+        }
+
+        #[allow(clippy::if_same_then_else)]
+        if amount_str.matches('.').count() > 1 {
+            // only 1 decimal point allowed
+            None
+        } else if amount_str.matches('-').count() > 1 {
+            None
+        } else if amount_str.contains('-') && !amount_str.starts_with('-') {
+            None
+        } else if amount_str.matches('.').count() == 0 {
+            // if there is no decimal point, then just add N zeros to the right and we're done
+            let zeros = "0".repeat(decimals);
+            let amount_str = amount_str + &zeros;
+
+            amount_str.parse::<SignedIntType>().ok().map(|v| SignedAmount { val: v })
+        } else {
```

Comment:
`.` and `-` are parsed into `None` but `-.` is parsed into `Some(SignedAmount { val: 0 })`. Is this correct?

Reply:
That is, when the number of decimals is more than zero

Reply:
I updated the implementation and added the tests. I think that's not the case anymore.